cylc/flow/host_select.py

```diff
@@ -0,0 +1,532 @@
+            compile(node, '<string>', 'eval'),
+            {'__builtins__': None},
+            variables
+        )
+    except Exception:
+        raise ValueError(expr)
+
+
+def _get_thresholds(string):
+    """Yield parsed threshold expressions.
+
+    Examples:
+        The first ``token.NAME`` encountered is returned as the query:
+        >>> _get_thresholds('foo() == 123').__next__()
+        (('foo',), 'RESULT == 123')
+
+        If multiple are present they will not get parsed:
+        >>> _get_thresholds('foo() in bar()').__next__()
+        (('foo',), 'RESULT in bar()')
+
+        Positional arguments are added to the query tuple:
+        >>> _get_thresholds('1 in foo("a")').__next__()
+        (('foo', 'a'), '1 in RESULT')
+
+        Comments (not in-line) and multi-line strings are permitted:
+        >>> _get_thresholds('''
+        ...     # earl of sandwhich
+        ...     foo() == 123
+        ...     # beef wellington
+        ... ''').__next__()
+        (('foo',), 'RESULT == 123')
+
+    Yields:
+        tuple - (query, expression)
+        query (tuple):
+            The method to call followed by any positional arguments.
+        expression (str):
+            The expression with the method call replaced by `RESULT`
+
+    """
+    for line in string.splitlines():
+        # parse the string one line at a time
+        # purposfully don't support multi-line expressions
+        line = line.strip()
+
+        if not line or line.startswith('#'):
+            # skip blank lines
+            continue
+
+        query = []
+        start = None
+        in_args = False
+
+        line_feed = BytesIO(line.encode())
+        for item in tokenize(line_feed.readline):
+            if item.type == token.ENCODING:
+                # encoding tag, not of interest
+                pass
+            elif not query:
+                # the first token.NAME has not yet been encountered
+                if item.type == token.NAME and item.string != 'in':
+                    # this is the first token.NAME, assume it it the method
+                    start = item.start[1]
+                    query.append(item.string)
+            elif item.string == '(':
+                # positional arguments follow this
+                in_args = True
+            elif item.string == ')':
+                # end of positional arguments
+                in_args = False
+                break
+            elif in_args:
+                # literal eval each argument
+                query.append(ast.literal_eval(item.string))
+        end = item.end[1]
+
+        yield (
+            tuple(query),
+            line[:start] + 'RESULT' + line[end:]
+        )
+
+
+@lru_cache()
+def _tuple_factory(name, params):
+    """Wrapper to namedtuple which cashes results to prevent duplicates."""
+    return namedtuple(name, params)
+
+
+def _deserialise(metrics, data):
+    """Convert dict to named tuples.
+
+    Examples:
+        >>> _deserialise(
```

Comment:
I think this would be easier to understand if you at least used words associated with quantities for metrics.

Reply:
This method is a generic data transformation, that data you pass into it doesn't really matter, it's not just for metrics.